Terminal.Gui/Core/KeyBindings.cs

```diff
@@ -0,0 +1,370 @@
+		/// </summary>
+		public Key OutKey { get; set; }
+		/// <summary>
+		/// Gets or sets the description for this key.
+		/// </summary>
+		public string Description { get; set; }
+		/// <summary>
+		/// Gets or sets the enable status.
+		/// </summary>
+		public bool Enabled { get; set; } = true;
+
+	}
+
+	/// <summary>
+	/// Manage all the <see cref="KeyBinding"/> collection used by the <see cref="Application.KeyBindings"/>
+	/// </summary>
+	public sealed class KeyBindings {
+		private class KeyBindingEqualityComparer : IEqualityComparer<KeyBinding> {
+			public bool Equals (KeyBinding x, KeyBinding y)
+			{
+				if (x.View == y.View && (x.InKey == y.InKey || x.OutKey == y.OutKey))
+					return true;
+				return false;
+			}
+
+			public bool ReferenceEquals (KeyBinding x, KeyBinding y)
+			{
+				if (x.View == y.View && x.InKey == y.InKey && x.OutKey == y.OutKey && x.Description == y.Description && x.Enabled == y.Enabled)
+					return true;
+				return false;
+			}
+
+			public int GetHashCode (KeyBinding obj)
+			{
+				if (obj == null)
+					throw new ArgumentNullException ();
+
+				int hCode = 0;
+				foreach (var c in obj.View) {
+					hCode += c;
+				}
+				return hCode.GetHashCode ();
+			}
+		}
+
+		/// <summary>
+		/// Initializes a default instance.
+		/// </summary>
+		public KeyBindings ()
+		{
+		}
+
+		/// <summary>
+		/// Initializes a instance with the necessary settings.
+		/// </summary>
+		/// <param name="view">The view.</param>
+		/// <param name="inKey">The input key pressed.</param>
+		/// <param name="outKey">The desired output key.</param>
+		/// <param name="description">The description.</param>
+		/// <param name="enabled"><c>true</c> if enabled, <c>false</c>otherwise.</param>
+		public KeyBindings (Type view, Key inKey, Key outKey, string description = "", bool enabled = true)
+		{
+			AddKey (view, inKey, outKey, description, enabled);
+		}
+
+		/// <summary>
+		/// Initializes a instance of a <see cref="KeyBinding"/> instance.
+		/// </summary>
+		/// <param name="keyBinding">The key binding.</param>
+		public KeyBindings (KeyBinding keyBinding)
+		{
+			AddKey (keyBinding);
+		}
+
+		/// <summary>
+		/// The views name with the enabled status.
+		/// </summary>
+		public List<(string View, bool Enabled)> Views { get; private set; } = new List<(string View, bool Enabled)> ();
+
+		/// <summary>
+		/// The <see cref="KeyBinding"/> collection.
+		/// </summary>
+		public List<KeyBinding> Keys { get; private set; } = new List<KeyBinding> ();
+
+		/// <summary>
+		/// Get or sets the enabled status.
+		/// </summary>
+		public bool Enabled { get; set; } = false;
+
+		/// <summary>
+		/// Get or sets the enable key to enter insert mode leaving the keys bindings.
+		/// </summary>
+		public Key EnableKey { get; set; } = Key.Enter;
+
+		/// <summary>
+		/// Gets or sets the disable key to leave insert mode returning to the keys bindings.
+		/// </summary>
+		public Key DisableKey { get; set; } = Key.Esc;
```

Comment:
I don't think we should set this out of the box.  Disabling keybindings on Enter/Esc would typically be unexpected for most programmers encountering this feature (the keybinding feature).  I would keep `bool Enabled` but leave it up to the API user to decide whether to use the feature and how to enable/disable it (what shortcut or menu item etc)

Reply:
Do you mean set the `bool Enabled = true` and get rid of the `Key EnableKey\Key DisableKey`?

Reply:
Yes :) I think that makes sense do you?

Reply:
More and less :-) The `bool Enabled = true` has no problem if there are no any key bindings setting but will force the `ProcessKey` in the `View` class to do a search in the key bindings unnecessarily, so the reason to setting as false.
For the `Key EnableKey\Key DisableKey` imagine if the user keyboard has malfunction on the `Enter` or `Esc` key and he wants to change these keys temporarily until to repair.

Reply:
Ah sorry I didn't mean that it should always be `Enter`/`Esc`

What I meant was to remove the functionality where a key (e.g. Enter) toggles the keybindings `Enabled` and instead leave it up to the API user to set that up themselves (e.g. with a status bar item or menu shortcut).

Since alternate keybindings have to be setup by the programmer in their program.  Disabling those keybindings should be quite a rare thing to do and shouldn't be tied to a commonly used button like `Enter`/`Esc`.

Even in the case of the vim keybindings you only want to disable the keybindings (`hjkl`) under certain scenarios not every time Enter is pressed and it might vary from control to control.

Reply:
But I'm using this functionality in the `ProcessKey` method on the `View` class. Sorry I forgot to remove the tied `Enter` and `Esc` from the `View` class. See the `UICatalog` sample where I binding the same keys for the `View` (which is a global binding) and for the `TextView`. In the `ListView` you can navigate with the `hjkl`, pressing `Enter` will leave the key binding. Try open the `Editor` and you can still navigate with `hjkl`. If the user want to use other keys to enter and leave key binding, he only need to change the `EnableKey` and `DisableKey` keys. In the `ProcessKey` method I changed the `Focused` to `MostFocused` property because is the view that has priority to run.